### PR TITLE
Extract new password in passwd_s (#246)

### PR DESCRIPTION
**Pull request by mistotebe**
Care has been taken to preserve backwards compatibility for the regular case of no password being requested too.

### CODE REVIEW

**Comment by mistotebe** [comment]
So you want me to add a `PASSWD_NEW = object()` into ldap.py?

Yes, [RFC 3062](https://tools.ietf.org/html/rfc3062#section-2.2) says so. Also, there is some confusion on the OPTIONAL status of genPasswd in the ASN.1 spec vs. text spec but I replicated the more permissive version for better interoperability.

**Comment by mistotebe** [comment]
@tiran Updated merge request, just the new_pw handling to resolve then?

**Comment by mistotebe** [comment]
Wouldn't your suggestion make it impossible to ever deprecate the old behaviour?

**Comment by mistotebe** [comment]
@tiran Ok, this might be better? If it is, feel free to squash the relevant commits before merging.

**Comment by encukou** [comment]
@tiran, do you want to finish the review?

**Comment by encukou** [comment]
My review will be sub-par, but it looks like I'll need to do it to move this forward. I'll dedicate a day next week to python-ldap.

**Comment by encukou** [comment]
@mistotebe Overall, this looks good to me!
I've rebased so tests can pass, and made a few nitpicks directly in your branch. Does that look OK?

The `extract_newpw` flag is not ideal API, but it should work for now. When https://github.com/python-ldap/python-ldap/issues/61 is implemented, it can be deprecated.

**Comment by mistotebe** [comment]
Thanks for the changes. I agree the result API needs a complete overhaul, couldn't do that here and frankly, the exception work is a prerequisite to that anyway.

File: Tests/t_ldapobject.py [hunk 1]
```diff
@@ -673,6 +673,45 @@ def test_compare_s_invalidattr(self):
...
+            dn,
+            [
+                ('objectClass', b'person'),
+                ('sn', b'PasswordTest'),
+                ('cn', b'PasswordTest'),
+                ('userPassword', b'initial'),
+            ]
+        )
+        self.assertEqual(result, ldap.RES_ADD)
+        self.assertEqual(type(msgid), type(0))
```
Comment (tiran) [inline Tests/t_ldapobject.py:691]:
```suggestion
        self.assertIsInstance(msgid, int)
```

File: Tests/t_ldapobject.py [hunk 2]
```diff
@@ -673,6 +673,45 @@ def test_compare_s_invalidattr(self):
...
+        # try changing password with a wrong old-pw
+        with self.assertRaises(ldap.UNWILLING_TO_PERFORM):
+            l.passwd_s(dn, "bogus", "ignored")
+
+        # have the server generate a new random pw
+        respoid, respvalue = l.passwd_s(dn, "initial", None, extract_newpw=True)
+        self.assertEqual(respoid, None)
+
+        password = respvalue.genPasswd
+        self.assertEqual(type(password), bytes)
```
Comment (tiran) [inline Tests/t_ldapobject.py:704]:
```suggestion
        self.assertIsInstance(password, bytes)
```

File: Lib/ldap/extop/__init__.py [hunk 3]
```diff
@@ -65,3 +65,4 @@ def decodeResponseValue(self,value):
 
 # Import sub-modules
 from ldap.extop.dds import *
+from ldap.extop.passwd import *
```
Comment (tiran) [inline Lib/ldap/extop/__init__.py:68]:
```suggestion
from ldap.extop.passwd import PasswordModifyResponse
```

**Comment by quanah** [comment]
> For some reason, Travis CI is not reporting results.

You may wish to see if the discussion here is of help: https://mta.openssl.org/pipermail/openssl-project/2020-June/002033.html

**Pull request merged**